chain/chain/src/resharding/resharding_actor.rs

```diff
@@ -1,23 +1,77 @@
-use super::types::FlatStorageSplitShardRequest;
-use near_async::messaging::{self, Handler};
+use super::types::ReshardingRequest;
+use crate::flat_storage_resharder::{FlatStorageResharder, FlatStorageReshardingTaskStatus};
+use crate::ChainStore;
+use near_async::futures::{DelayedActionRunner, DelayedActionRunnerExt};
+use near_async::messaging::{self, HandlerWithContext};
+use near_primitives::hash::CryptoHash;
+use near_primitives::types::BlockHeight;
+use near_store::{ShardUId, Store};
+use time::Duration;
 
 /// Dedicated actor for resharding V3.
-pub struct ReshardingActor {}
+pub struct ReshardingActor {
+    chain_store: ChainStore,
+}
 
 impl messaging::Actor for ReshardingActor {}
 
-impl Handler<FlatStorageSplitShardRequest> for ReshardingActor {
-    fn handle(&mut self, msg: FlatStorageSplitShardRequest) {
-        self.handle_flat_storage_split_shard_request(msg);
+impl HandlerWithContext<ReshardingRequest> for ReshardingActor {
+    fn handle(&mut self, msg: ReshardingRequest, ctx: &mut dyn DelayedActionRunner<Self>) {
+        match msg {
+            ReshardingRequest::FlatStorageSplitShard { resharder } => {
+                self.handle_flat_storage_split_shard(resharder);
+            }
+            ReshardingRequest::FlatStorageShardCatchup {
+                resharder,
+                shard_uid,
+                flat_head_block_hash,
+            } => {
+                // Shard catchup task is delayed and could get postponed several times. This must be
+                // done to cover the scenario in which catchup is triggered so fast that the initial
+                // state of the new flat storage is beyond the chain final tip.
+                ctx.run_later(
```

Comment:
Maybe `run_later` is misused here. This has been my tentative solution to the following problem.

Originally, I was calling directly `handle_flat_storage_shard_catchup` and waiting some time inside the function. But it was failing in test loop for resharding_v3 because the test was getting stuck inside this function call. The other actors weren't progressing at all and the condition to resume catchup wasn't satisfied ever.

Basically I need a way to postpone execution of catchup until the canonical chain makes enough progress. I found that `run_later` achieves that by allowing other actors to make progress in the meantime. 

Should I rather change the test_loop test to avoid this problem?
Please feel free to suggest a better approach (even offline)

Reply:
[This](https://github.com/near/nearcore/blob/e6dd1a447a950ae3716b9d682e3c305cf715d86b/chain/client/src/sync_jobs_actor.rs) is an example of how run_later is used. We can discuss this offline